### PR TITLE
Pluggable Allocators: tcmalloc support for OE SDK

### PR DESCRIPTION
**Pull request by anakrish**
Lockdown tcmalloc (version 2.7) support for SGX enclaves.

This is based on SGX SDK for Linux's support for tcmalloc: 
https://github.com/intel/linux-sgx/tree/master/sdk/gperftools

tcmalloc is ported to work within enclaves at: 
https://github.com/openenclave/gperftools/tree/openenclave-port-2.7

tcmalloc's openenclave fork is added as a submodule.
Pluggability is locked down via test/tcmalloc.
Additionally threading tests are also locked down by plugging in oetcmalloc.

Signed-off-by: Anand Krishnamoorthi <anakrish@microsoft.com>

### CODE REVIEW

**Comment by anakrish** [comment]
bors try

**Comment by mingweishih** [comment]
> Lockdown tcmalloc (version 2.7) support for SGX enclaves.
> 
> This is based on SGX SDK for Linux's support for tcmalloc:
> https://github.com/intel/linux-sgx/tree/master/sdk/gperftools
> 
> tcmalloc is ported to work within enclaves at:
> https://github.com/openenclave/gperftools/tree/openenclave-port-2.7
> 
> tcmalloc's openenclave fork is added as a submodule.
> Pluggability is locked down via test/tcmalloc.
> Additionally threading tests are also locked down by plugging in oetcmalloc.
> 
> Signed-off-by: Anand Krishnamoorthi [anakrish@microsoft.com](mailto:anakrish@microsoft.com)

Is there a specific reason for adding the threading tests for tcmalloc? What about snmalloc?

File: tests/thread/enc/CMakeLists.txt [hunk 1]
```diff
@@ -43,3 +43,43 @@ enclave_include_directories(pthread_enc PRIVATE ${CMAKE_CURRENT_BINARY_DIR}
...
+  pthread_enc_tcmalloc PRIVATE ${CMAKE_CURRENT_BINARY_DIR}
+  ${CMAKE_CURRENT_SOURCE_DIR})
+
+enclave_link_libraries(pthread_enc_tcmalloc oetcmalloc)
+
+enclave_include_directories(
+  oethread_enc_tcmalloc PRIVATE ${CMAKE_CURRENT_BINARY_DIR}
+  ${CMAKE_CURRENT_SOURCE_DIR})
+
+enclave_link_libraries(oethread_enc_tcmalloc oetcmalloc)
```
Comment (mingweishih) [inline tests/thread/enc/CMakeLists.txt:85]: Is the design for user to choose which allocator to use (i.e., specify the library to link)?

File: 3rdparty/CMakeLists.txt [hunk 2]
```diff
@@ -24,6 +24,10 @@ if (CODE_COVERAGE OR BUILD_LIBGCOV)
   add_subdirectory(libgcov)
 endif ()
 
+if (OE_SGX)
+  add_subdirectory(tcmalloc)
```
[the remark this answers is not in the document]
Comment (anakrish) [reply 3rdparty/CMakeLists.txt:28]: snmalloc has already been locked down via tests/snmalloc. It does not however compile with GCC 5.4.0 which is default for Ubuntu 16.04. Hence we do this:
https://github.com/openenclave/openenclave/blob/df6fa6ef1472fc04fe68f7a2f9e495791131b571/3rdparty/CMakeLists.txt#L17-L21

File: tests/thread/enc/CMakeLists.txt (hunk 1 above)
Comment (anakrish) [reply tests/thread/enc/CMakeLists.txt:85]: > Is the design for user to choose which allocator to use (i.e., specify the library to link)?

Yes, the user plugs-in an allocator just be specifying it in the linker line. This is similar to how GBLIC's malloc is replaced.

This PR just locks down the support for tcmalloc; we don't make it available to the user.

Currently, there are two choices in which we can make allocators (snmalloc, tcmalloc) available to the user
- Create an OE SDK sample that shows how to plug in both the allocators (and hence any other allocator that has been ported to work within enclaves).
- Create a separate repository where we show how allocators are plugged in. This has the advantage that more allocators could be supported in the future and could be added without creating more submodule dependencies in OE repository.

In both the above approaches, we just show pluggability; as an advanced feature.

**Comment by anakrish** [comment]
> > Lockdown tcmalloc (version 2.7) support for SGX enclaves.
> > This is based on SGX SDK for Linux's support for tcmalloc:
> > https://github.com/intel/linux-sgx/tree/master/sdk/gperftools
> > tcmalloc is ported to work within enclaves at:
> > https://github.com/openenclave/gperftools/tree/openenclave-port-2.7
> > tcmalloc's openenclave fork is added as a submodule.
> > Pluggability is locked down via test/tcmalloc.
> > Additionally threading tests are also locked down by plugging in oetcmalloc.
> > Signed-off-by: Anand Krishnamoorthi [anakrish@microsoft.com](mailto:anakrish@microsoft.com)
> 
> Is there a specific reason for adding the threading tests for tcmalloc? What about snmalloc?

Yes. High-performant allocators like snmalloc, tcmalloc are most useful in multi-threaded scenarios.

tcmalloc has a fixed overhead (0x900000 bytes according to [Intel SDK](https://github.com/intel/linux-sgx/tree/master/sdk/gperftools)). We can easily run the multi-threaded tests with tcmalloc.

snmalloc currently has high memory requirements. The thread-tests would crash at runtime due to overhead of the allocator though the test itself does not consume much memory.

**Comment by dthaler** [comment]
I see that tcmalloc.edl  imports the optee platform.edl too but other code seems to check for if SGX, which seems inconsistent.  Is there anything actually SGX-only in tcmalloc?

**Comment by anakrish** [comment]
> I see that tcmalloc.edl imports the optee platform.edl too but other code seems to check for if SGX, which seems inconsistent. Is there anything actually SGX-only in tcmalloc?

As far as I can tell, tcmalloc can be compiled for ARM and can therefore be made to work for OP-TEE.
However, currently OP-TEE does not support multiple threads within the enclave and therefore not much benefit can be gained by using tcmalloc or other multi-thread optimized allocators.

File: 3rdparty/tcmalloc/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+check_submodule_not_empty("${PROJECT_SOURCE_DIR}/3rdparty/tcmalloc/gperftools")
+
+if (NOT OE_SGX)
+  return()
+endif ()
+
+# oesnmalloc depends on musl_includes and libcxx_includes.
+# However specifying those external projects as dependencies does not seem to
+# work. Build of oesnmalloc is launched while musl_includes is being built.
```
Comment (achamayou) [inline 3rdparty/tcmalloc/CMakeLists.txt:12]:
```suggestion
# work. Build of oetcmalloc is launched while musl_includes is being built.
```

File: 3rdparty/tcmalloc/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+# This likely a cmake/ninja bug. Depending on oelibc_includes (which is an
+# INTERFACE library that depends on musl_includes) also does not work.
+# To ensure that musl_includes and libcxx_includes external projects are
+# completed before oesnmalloc is built, we use the custom command below
```
Comment (achamayou) [inline 3rdparty/tcmalloc/CMakeLists.txt:16]:
```suggestion
# completed before oetcmalloc is built, we use the custom command below
```

File: 3rdparty/tcmalloc/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+# that depends on the external projects. The custom command copies header
+# files generated by the external projects to oesnmalloc's binary directory.
```
Comment (achamayou) [inline 3rdparty/tcmalloc/CMakeLists.txt:18]:
```suggestion
# files generated by the external projects to oetcmalloc's binary directory.
```

File: tests/tcmalloc/enc/enc.c [hunk 6]
```diff
@@ -0,0 +1,80 @@
+// Copyright (c) Open Enclave SDK contributors.
+// Licensed under the MIT License.
+
+#include <openenclave/internal/tests.h>
+#include <stdlib.h>
+
+#include "tcmalloc_t.h"
+
+// snmalloc requires at least 4K heap pages.
```
Comment (achamayou) [inline tests/tcmalloc/enc/enc.c:9]: This is no longer accurate for snmalloc, and doesn't apply to tcmalloc?

**Comment by achamayou** [comment]
@anakrish as discussed separately, snmalloc's initial requirements can and have been adjusted, so the current requirements when using the OE PAL are 64 + 64 * Nthreads, which is rather less than 3 * 1024 pages. I think it is not comparatively a "high memory requirement" :)

**Comment by xiaonan-INTC** [comment]
Hi, @anakrish 

This PR has failed for some days and it seems no one is fixing it, could you help check?
I have some test cases and will check if they could be ported to OE SDK after it is ready.

**Comment by radhikaj** [comment]
Closing this for now as this is not pressing. We can reopen when we need to.

**Pull request closed without merging**